js/src/SearchBar.js

```diff
@@ -1,158 +1,171 @@
+        if (r.type === 'reaction') {
+          this.props.map.zoom_to_reaction(r.reaction_id)
+          this.props.map.highlight_reaction(r.reaction_id)
+        } else if (r.type === 'metabolite') {
+          this.props.map.zoom_to_node(r.node_id)
+          this.props.map.highlight_node(r.node_id)
+        } else if (r.type === 'text_label') {
+          this.props.map.zoom_to_text_label(r.text_label_id)
+          this.props.map.highlight_text_label(r.text_label_id)
+        } else {
+          throw new Error('Bad search index data type: ' + r.type)
+        }
+      }
+      this.setState({
+        searchItem: value,
+        current: 1,
+        counter,
+        results
+      })
+    }
+  }
 
-function is_visible() {
-    return this.selection.style('display') !== 'none';
-}
+  dropDuplicates (results) {
+    const compKeys = {
+      metabolite: ['m', 'node_id'],
+      reaction: ['r', 'reaction_id'],
+      text_label: ['t', 'text_label_id']
+    }
+    return _.uniq(results, item => {
+      // make a string for fast comparison
+      var t = compKeys[item.type]
+      return t[0] + item[t[1]]
+    })
+  }
 
-function toggle(on_off) {
-    if (on_off===undefined) this.is_active = !this.is_active;
-    else this.is_active = on_off;
+  next () {
+    if (this.state.results) {
+      if (this.state.current === this.state.results.length) {
+        this.update(1)
+      } else {
+        this.update(this.state.current + 1)
+      }
+    }
+  }
 
-    if (this.is_active) {
-        this.selection.style('display', null);
-        this.counter.text('');
-        this.input.node().value = '';
-        this.input.node().focus();
-        // escape key
-        this.clear_escape = this.map.key_manager
-            .add_escape_listener(function() {
-                this.toggle(false);
-            }.bind(this), true);
-        // next keys
-        this.clear_next = this.map.key_manager
-            .add_key_listener(['enter', 'ctrl+g'], function() {
-                this.next();
-            }.bind(this), false);
-        // previous keys
-        this.clear_previous = this.map.key_manager
-            .add_key_listener(['shift+enter', 'shift+ctrl+g'], function() {
-                this.previous();
-            }.bind(this), false);
-        // run the show callback
-        this.callback_manager.run('show');
-    } else {
-        this.map.highlight(null);
-        this.selection.style('display', 'none');
-        this.results = null;
-        if (this.clear_escape) this.clear_escape();
-        this.clear_escape = null;
-        if (this.clear_next) this.clear_next();
-        this.clear_next = null;
-        if (this.clear_previous) this.clear_previous();
-        this.clear_previous = null;
-        // run the hide callback
-        this.callback_manager.run('hide');
+  previous () {
+    if (this.state.results) {
+      if (this.state.current === 1) {
+        this.update(this.state.results.length)
```

Comment:
Most of the comments I wrote for `next` apply here to.
If indexing would go from `0...this.state.results.length-1`, this could be
```
this.update((this.state.current + this.state.results.length -1) % this.state.results.length)
```
The two logic could be merged.